server/src/main/java/org/apache/lucene/index/LazySoftDeletesDirectoryReaderWrapper.java

```diff
@@ -0,0 +1,288 @@
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import org.apache.lucene.document.Field;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.DocValuesFieldExistsQuery;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.FixedBitSet;
+import org.elasticsearch.common.lucene.Lucene;
+
+/**
+ * This is a modified version of {@link SoftDeletesDirectoryReaderWrapper} that materializes the liveDocs
+ * bitset lazily. In contrast to {@link SoftDeletesDirectoryReaderWrapper}, this wrapper can only be used
+ * for non-NRT readers.
+ *
+ * This reader filters out documents that have a doc values value in the given field and treat these
+ * documents as soft deleted. Hard deleted documents will also be filtered out in the live docs of this reader.
+ * @see IndexWriterConfig#setSoftDeletesField(String)
+ * @see IndexWriter#softUpdateDocument(Term, Iterable, Field...)
+ * @see SoftDeletesRetentionMergePolicy
+ */
+public final class LazySoftDeletesDirectoryReaderWrapper extends FilterDirectoryReader {
+    private final String field;
+    private final CacheHelper readerCacheHelper;
+    /**
+     * Creates a new soft deletes wrapper.
+     * @param in the incoming directory reader
+     * @param field the soft deletes field
+     */
+    public LazySoftDeletesDirectoryReaderWrapper(DirectoryReader in, String field) throws IOException {
+        this(in, new LazySoftDeletesSubReaderWrapper(Collections.emptyMap(), field));
+    }
+
+    private LazySoftDeletesDirectoryReaderWrapper(DirectoryReader in, LazySoftDeletesSubReaderWrapper wrapper) throws IOException {
+        super(in, wrapper);
+        this.field = wrapper.field;
+        readerCacheHelper = in.getReaderCacheHelper() == null ? null : new DelegatingCacheHelper(in.getReaderCacheHelper());
+    }
+
+    @Override
+    protected DirectoryReader doWrapDirectoryReader(DirectoryReader in) throws IOException {
+        Map<CacheKey, LeafReader> readerCache = new HashMap<>();
+        for (LeafReader reader : getSequentialSubReaders()) {
+            // we try to reuse the live docs instances here if the reader cache key didn't change
+            if (reader instanceof LazySoftDeletesFilterCodecReader && reader.getReaderCacheHelper() != null) {
+                readerCache.put(((LazySoftDeletesFilterCodecReader) reader).reader.getReaderCacheHelper().getKey(), reader);
+            }
+
+        }
+        return new LazySoftDeletesDirectoryReaderWrapper(in, new LazySoftDeletesSubReaderWrapper(readerCache, field));
+    }
+
+    @Override
+    public CacheHelper getReaderCacheHelper() {
+        return readerCacheHelper;
+    }
+
+    private static class LazySoftDeletesSubReaderWrapper extends SubReaderWrapper {
+        private final Map<CacheKey, LeafReader> mapping;
```

Comment:
We should be able to drop this mapping too, which is only relevant for readers that may be re-opened.

Reply:
fixed in ba54eb3